includes/AMP/Canonical_Sanitizer.php

```diff
@@ -54,28 +54,43 @@ class Canonical_Sanitizer extends AMP_Base_Sanitizer {
 	 * @return void
 	 */
 	public function sanitize() {
-		$links         = [];
+		/**
+		 * The rel=canonical DOM element.
+		 *
+		 * @var DOMElement The element.
+		 */
+		$rel_canonical = null;
 		$link_elements = $this->dom->head->getElementsByTagName( Tag::LINK );
 
+		// This fallback to get_permalink() ensures that there's a canonical link
+		// even when previewing drafts.
+		$canonical_url = wp_get_canonical_url();
+		if ( ! $canonical_url ) {
+			$canonical_url = get_permalink();
+		}
+
 		/**
 		 * Link element.
 		 *
 		 * @var DOMElement $link
 		 */
 		foreach ( $link_elements as $link ) {
-			if ( $link->hasAttribute( Attribute::REL ) ) {
-				$links[ $link->getAttribute( Attribute::REL ) ][] = $link;
+			if ( $link->getAttribute( Attribute::REL ) === Attribute::REL_CANONICAL ) {
+				if ( ! empty( $link->getAttribute( Attribute::HREF ) ) ) {
+					return;
+				}
+
+				$link->setAttribute( Attribute::HREF, (string) $canonical_url );
+
+				break;
 			}
```

Comment:
Why does this need to be a foreach loop?

Reply:
It doesn't _need_ to be, but it was a loop already before and seems readable to me.

What would be an alternative? An XPath query?

Reply:
Right, I think that might be slightly better:

```php
// Ensure link[rel=canonical] is in head.
$rel_canonical = $this->dom->xpath->query( './link[rel=canonical]', $this->dom->head )->item( 0 );
if ( ! $rel_canonical instanceof DOMElement ) {
	$rel_canonical = $this->dom->createElement( Tag::LINK );
	$rel_canonical->setAttribute( Attribute::REL, Attribute::REL_CANONICAL );
	$this->dom->head->appendChild( $rel_canonical );
}

// Ensure link[rel=canonical] has a non-empty href attribute.
if ( empty( $rel_canonical->getAttribute( Attribute::HREF ) ) ) {
	$rel_canonical->setAttribute( Attribute::HREF, (string) $canonical_url );
}
```